File: ios/CalculationsModule.swift
import Foundation

@objc(CalculationsModule)
class CalculationsModule: NSObject {
  
  @objc
  func performIntensiveCalculations(_ iterations: Int, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
      DispatchQueue.global(qos: .background).async {
          do {
              let results = self.performIntensiveCalculations(iterations)
              resolve(results)
          } catch let error {
              reject("Error", error.localizedDescription, error)
          }
      }
  }
  
  @objc static func requiresMainQueueSetup() -> Bool {
         return false
     }
  
  private func calculateFibonacci(_ num: Int) -> Int {
          if num <= 1 { return num }
          return calculateFibonacci(num - 1) + calculateFibonacci(num - 2)
      }
  
  
  private func performIntensiveCalculations(_ iterations: Int) -> [Int] {
          var results: [Int] = []
          for _ in 0..<iterations {
              let result = calculateFibonacci(40) // Simulate intensive calculation
              results.append(result)
          }
          return results
      }
}

File: android/app/src/main/java/com/nativecodedemo/CalculationsModule.kt
package com.nativecodedemo

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import java.util.concurrent.Executors

class CalculationsModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    private val executor = Executors.newSingleThreadExecutor()


    override fun getName() = "CalculationsModule"


    @ReactMethod
    fun performIntensiveCalculations(iterations: Int, promise: Promise) {
        executor.execute {
            try {
                val results = performIntensiveCalculations(iterations)
                promise.resolve(Arguments.fromList(results))
            } catch (e: Exception) {
                promise.reject("Error", e)
            }
        }
    }
    private fun calculateFibonacci(num: Int): Int {
        if (num <= 1) return num
        return calculateFibonacci(num - 1) + calculateFibonacci(num - 2)
    }

    private fun performIntensiveCalculations(iterations: Int): List<Int> {
        val results = mutableListOf<Int>()
        for (i in 0 until iterations) {
            val result = calculateFibonacci(40) // High enough number to simulate intensive calculation
            results.add(result)
        }
        return results
    }

}